### ISSUE TITLE
indexer fails with 'Too many open files' since 1.1-rc20

### ISSUE THREAD

**Issue by vladak** [oracle/opengrok#1993]
Since 1.1-rc20 the indexing of "prepped" (i.e. unpacked with patches applied) version of https://github.com/oracle/solaris-userland gives me:

```
2018-01-29 09:03:01.638+0100 SEVERE t37 Indexer$1.run: An error occured while updating index
java.nio.file.FileSystemException: /opengrok/data/index/userland-default-prepped: Too many open files
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:91)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
        at sun.nio.fs.UnixFileSystemProvider.newDirectoryStream(UnixFileSystemProvider.java:427)
        at java.nio.file.Files.newDirectoryStream(Files.java:457)
        at org.apache.lucene.store.FSDirectory.listAll(FSDirectory.java:215)
        at org.apache.lucene.store.FSDirectory.listAll(FSDirectory.java:234)
        at org.apache.lucene.store.FilterDirectory.listAll(FilterDirectory.java:57)
        at org.apache.lucene.index.IndexFileDeleter.refresh(IndexFileDeleter.java:429)
        at org.apache.lucene.index.IndexWriter.rollbackInternalNoCommit(IndexWriter.java:2249)
        at org.apache.lucene.index.IndexWriter.rollbackInternal(IndexWriter.java:2193)
        at org.apache.lucene.index.IndexWriter.shutdown(IndexWriter.java:1129)
        at org.apache.lucene.index.IndexWriter.close(IndexWriter.java:1171)
        at org.opensolaris.opengrok.index.IndexDatabase.update(IndexDatabase.java:487)
        at org.opensolaris.opengrok.index.Indexer$1.run(Indexer.java:1046)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

Either it became more demanding (I think the default per process limit is 256 fd's) or there is file descriptor leak.

**Comment by vladak** [comment]
Until this is cleared I am marking this as a bug.

**Comment by tulinkry** [comment]
it can be connected to the parallelization but the pool of all together processed files is controled with thread count I think. So this would lead to fd leak. @idodeclare ?

**Comment by vladak** [comment]
There's no doubt this happened with the transition from rc19 to rc20. I deployed rc20 on Friday around 3 PM and the first indexing that ran at 4 PM contained the exception and almost all indexer runs ever since (15 out of 16).

**Comment by vladak** [comment]
For completeness, this was per-project indexing (`indexpart` used in the `OpenGrok` script).

**Comment by vladak** [comment]
Today the `EMFILE` error happened elsewhere, interestingly in the index merge operation as mentioned in PR #1936:

```
2018-01-29 12:58:47.181+0100 WARNING t37 IndexDatabase.indexParallel: 3 failures (0.0%) while parallel-indexing
2018-01-29 12:59:04.750+0100 SEVERE t145 Executor$3.uncaughtException: Uncaught exception in thread Lucene Merge Thread #1 with ID 145: java.nio.file.FileSystemException: /opengrok/data/index/userland-default-prepped/_d5sib_Lucene54_0.dvm: Too many open files
org.apache.lucene.index.MergePolicy$MergeException: java.nio.file.FileSystemException: /opengrok/data/index/userland-default-prepped/_d5sib_Lucene54_0.dvm: Too many open files
        at org.apache.lucene.index.ConcurrentMergeScheduler.handleMergeException(ConcurrentMergeScheduler.java:703)
        at org.apache.lucene.index.ConcurrentMergeScheduler$MergeThread.run(ConcurrentMergeScheduler.java:683)
Caused by: java.nio.file.FileSystemException: /opengrok/data/index/userland-default-prepped/_d5sib_Lucene54_0.dvm: Too many open files
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:91)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
        at sun.nio.fs.UnixFileSystemProvider.newFileChannel(UnixFileSystemProvider.java:177)
        at java.nio.channels.FileChannel.open(FileChannel.java:287)
        at java.nio.channels.FileChannel.open(FileChannel.java:335)
        at org.apache.lucene.store.MMapDirectory.openInput(MMapDirectory.java:238)
        at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:99)
        at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:99)
        at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:99)
        at org.apache.lucene.store.Directory.openChecksumInput(Directory.java:137)
        at org.apache.lucene.codecs.lucene50.Lucene50CompoundFormat.write(Lucene50CompoundFormat.java:93)
        at org.apache.lucene.index.IndexWriter.createCompoundFile(IndexWriter.java:4945)
        at org.apache.lucene.index.IndexWriter.mergeMiddle(IndexWriter.java:4421)
        at org.apache.lucene.index.IndexWriter.merge(IndexWriter.java:3931)
        at org.apache.lucene.index.ConcurrentMergeScheduler.doMerge(ConcurrentMergeScheduler.java:624)
        at org.apache.lucene.index.ConcurrentMergeScheduler$MergeThread.run(ConcurrentMergeScheduler.java:661)

2018-01-29 12:59:04.751+0100 SEVERE t37 Indexer$1.run: An error occured while updating index
java.lang.IllegalStateException: this writer hit an unrecoverable error; cannot commit
        at org.apache.lucene.index.IndexWriter.prepareCommitInternal(IndexWriter.java:2986)
        at org.apache.lucene.index.IndexWriter.commitInternal(IndexWriter.java:3244)
        at org.apache.lucene.index.IndexWriter.shutdown(IndexWriter.java:1128)
        at org.apache.lucene.index.IndexWriter.close(IndexWriter.java:1171)
        at org.opensolaris.opengrok.index.IndexDatabase.update(IndexDatabase.java:487)
        at org.opensolaris.opengrok.index.Indexer$1.run(Indexer.java:1046)
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
        at java.util.concurrent.FutureTask.run(FutureTask.java:266)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.nio.file.FileSystemException: /opengrok/data/index/userland-default-prepped/_d5sib_Lucene54_0.dvm: Too many open files
        at sun.nio.fs.UnixException.translateToIOException(UnixException.java:91)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
        at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
        at sun.nio.fs.UnixFileSystemProvider.newFileChannel(UnixFileSystemProvider.java:177)
        at java.nio.channels.FileChannel.open(FileChannel.java:287)
        at java.nio.channels.FileChannel.open(FileChannel.java:335)
        at org.apache.lucene.store.MMapDirectory.openInput(MMapDirectory.java:238)
        at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:99)
        at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:99)
        at org.apache.lucene.store.FilterDirectory.openInput(FilterDirectory.java:99)
        at org.apache.lucene.store.Directory.openChecksumInput(Directory.java:137)
        at org.apache.lucene.codecs.lucene50.Lucene50CompoundFormat.write(Lucene50CompoundFormat.java:93)
        at org.apache.lucene.index.IndexWriter.createCompoundFile(IndexWriter.java:4945)
        at org.apache.lucene.index.IndexWriter.mergeMiddle(IndexWriter.java:4421)
        at org.apache.lucene.index.IndexWriter.merge(IndexWriter.java:3931)
        at org.apache.lucene.index.ConcurrentMergeScheduler.doMerge(ConcurrentMergeScheduler.java:624)
        at org.apache.lucene.index.ConcurrentMergeScheduler$MergeThread.run(ConcurrentMergeScheduler.java:661)
```

**Comment by idodeclare** [comment]
> So this would suggest fd leaks. (@tulinkry)

[File Leak Detector](http://file-leak-detector.kohsuke.org/) shows no bad usage patterns, and a periodic sampling of `lsof -p` for a run does not show unbounded growth of active, open descriptors.

> Either it became more demanding (I think the default per process limit is 256 fd's) or there is file descriptor leak. (@vladak)

It became more demanding in a couple of ways:

1) `indexDown()` now accumulates all the work to do before scheduling it, and `indexParallel()` then tackles it as quickly as it can on the `ForkJoinPool`. Previously, `addFile()` would have been called as the Lucene index was traversed, so there was a slight throttling there.

2) Lucene commits still happen on the executor pool not the `ForkJoinPool`. The executor pool is shrunken in rc20 (previously it was twice the number of processors), and those threads are clients of the `ForkJoinPool`. The number of threads active is unchanged (i.e. "twice the number of processors" with _n_ for the executor and _n_ for the `ForkJoinPool` but by design the `ForkJoinPool` is active nearly the entire time.

`ulimit -n` of 256 is a bit low for a process that now wants to actively use 32 processors to read and write files in OpenGrok's usage pattern, with six simultaneous descriptors used for each `addFile()`: two for `Ctags` pipes, two for `REFS` and `FULL` tokenizers' opening read streams, and two for the `writeXref()` input and output streams.

That leaves only 64 descriptors for all the JARs that the JVM has to keep open plus the Lucene writing threads running for each executor that is targeting a Lucene index.

**Comment by vladak** [comment]
Okay. The 256 default limit is definitely low. I think we should just try to bump the open files soft (current) limit to hard limit in the `OpenGrok` shell script - I am not aware of how to do the `setrlimit` equivalent in Java.

**Comment by vladak** [comment]
Plus a note in https://github.com/oracle/opengrok#9-tuning-opengrok-for-large-code-bases might be in order.

**Comment by vladak** [comment]
For reference, `ulimit -n 1024` was not enough in my case, using the recommended 10k worked :-)

**Comment by vladak** [comment]
Looks like 10k is still not enough - it worked previously because I was running the previous test without full "prepping" phase.

**Comment by vladak** [comment]
Tried 32k (confirmed with dtrace that `process.max-file-descriptor` was set to 32k upon `exece` syscall) and even that was not enough. Looks like this should be investigated in more depth.

**Comment by vladak** [comment]
I put a breakpoint to `writer.close()` in `update(IndexerParallelizer parallelizer)` and discovered with `pfiles` that there are thousands of xref files opened. First 32 fds is the usual stuff and then:

```
  32: S_IFSOCK mode:0666 dev:566,0 ino:51412 uid:0 gid:0 size:0
      O_RDWR
        SOCK_STREAM
        SO_SNDBUF(16384),SO_RCVBUF(5120)
        sockname: AF_UNIX 
        peer: java[14] zone: global[0]
  33: S_IFREG mode:0644 dev:216,65538 ino:281708 uid:0 gid:2 size:3860502
      O_RDONLY|O_LARGEFILE FD_CLOEXEC
      /usr/jdk/instances/jdk1.8.0/jre/lib/ext/cldrdata.jar
      offset:3860502
  34: S_IFDIR mode:0755 dev:216,65558 ino:35314580 uid:80 gid:80 size:3
      O_RDONLY|O_LARGEFILE
      /opengrok/data/xref/userland-default-prepped/components/python/nose/nose-1.3.7/build/tests/functional_tests/doc_tests/test_issue145/support/package1
      offset:353144430
  35: S_IFDIR mode:0755 dev:216,65558 ino:35314589 uid:80 gid:80 size:3
      O_RDONLY|O_LARGEFILE FD_CLOEXEC
      /opengrok/data/xref/userland-default-prepped/components/python/nose/nose-1.3.7/build/tests/functional_tests/doc_tests/test_issue145/support/package2c
      offset:372910834

...

32765: S_IFDIR mode:0755 dev:216,65558 ino:27408553 uid:80 gid:80 size:5
      O_RDONLY|O_LARGEFILE FD_CLOEXEC
      /opengrok/data/xref/userland-default-prepped/components/desktop/firefox/firefox-52.6.0esr/devtools/client/memory/test
      offset:464802953
32766: S_IFDIR mode:0755 dev:216,65558 ino:27408553 uid:80 gid:80 size:5
      O_RDONLY|O_LARGEFILE
      /opengrok/data/xref/userland-default-prepped/components/desktop/firefox/firefox-52.6.0esr/devtools/client/memory/test
      offset:464802953
```

What is special about them ? These are all directories. So there is a problem after all - this is not going to scale even if I bumped the limit to the 65k hard limit.

**Comment by vladak** [comment]
Interestingly https://github.com/kohsuke/file-leak-detector did not detect anything even when the exception was thrown or when threshold for number of opened fds was specified; had to use remote debugger to see what is going on.

**Comment by tulinkry** [comment]
https://github.com/oracle/opengrok/blob/master/src/org/opensolaris/opengrok/index/PendingFileCompleter.java#L318-L347 this code use a stream and does not close it and it should be just directories - can this be it?

**Comment by idodeclare** [comment]
> can this be it?

Good catch. That's supposed to be a try-with-resources statement.

I was looking to see if parallelized `mkdirs()` might leave a dangling fd — but you've found a believable culprit.

**Comment by idodeclare** [comment]
> Interestingly https://github.com/kohsuke/file-leak-detector did not detect anything even when the exception was thrown or when threshold for number of opened fds was specified ....

From the project's changelog, it does seem that the agent attaches to file-oriented classes and operations.

**Comment by vladak** [comment]
fixed in be85104
